Add Zoom and Compass widgets

ilyabo opened the pull request:
For https://github.com/visgl/deck.gl/issues/7946

### Change List

- Add zoom widgets which renders Zoom in and Zoom out buttons
- Add compass widget which resets bearing and pitch when clicked


![Screen Shot 2024-03-18 at 9 20 37 PM](https://github.com/visgl/deck.gl/assets/2461547/242a6e47-3b8c-458a-adc2-419b498c9b5d)

ibgreen commented on modules/widgets/src/zoom-widget.tsx:
```diff
@@ -0,0 +1,109 @@
+  viewId?: string | null;
+  placement?: WidgetPlacement;
+  /**
+   * A [compatible DOM element](https://developer.mozilla.org/en-US/docs/Web/API/Element/requestFullScreen#Compatible_elements) which should be made full screen.
```
Copy paste (full screen)

ilyabo replied:
Oops, will remove

ibgreen commented on modules/widgets/src/zoom-widget.tsx:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  onAdd({deck}: {deck: Deck}): HTMLDivElement {
+    const el = document.createElement('div');
```
Do we need to use abbreviations `el` => `elements`

Pessimistress commented on modules/widgets/src/zoom-widget.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  handleZoom(nextZoom: number) {
+    const viewId = this.viewId || 'default-view';
+    const nextViewState = {...this.viewport, zoom: nextZoom};
```
add smooth easing?

ibgreen commented on examples/get-started/pure-js/widgets/app.js:
```diff
@@ -67,5 +67,5 @@ new Deck({
     })
   ],
-  widgets: [new FullscreenWidget({style: widgetTheme})]
+  widgets: [new ZoomWidget({style: widgetTheme}), new FullscreenWidget({style: widgetTheme})]
```
Q: What happens if we provide these multiple times? Every render will create a new set of DOM elements?

chrisgervang replied:
Yes, and they stack vertically by default (`display: block;`).

ilyabo replied:
Yes, looks like this:

<img src=https://github.com/visgl/deck.gl/assets/351828/63977c70-c395-42e9-95c7-840ba2a0c354 width=300>

chrisgervang commented:
I did a cleanup pass on the [design file](https://www.figma.com/file/KGQqLbZZGUINgyyKYPvDvu/Widgets?type=design&node-id=419-1602&mode=dev), so I'd say it's a stable reference now. 

I was thinking a press and hold interaction would be nice for this widget so that users can have a smooth zoom over a long range. I've added this interaction to our https://www.figma.com/proto/KGQqLbZZGUINgyyKYPvDvu/Widgets?type=design&node-id=419-1643&t=x8lMA1JjZ6fRbtUF-0&scaling=min-zoom&page-id=230%3A839&starting-point-node-id=419%3A1643.

chrisgervang commented on modules/widgets/src/compass-widget.tsx:
```diff
@@ -0,0 +1,118 @@
+  id: string;
+  viewId?: string | null;
+  placement?: WidgetPlacement;
+  CompassLabel?: string;
```
This can be `label`

chrisgervang commented on modules/widgets/src/compass-widget.tsx:
```diff
@@ -0,0 +1,118 @@
+  style?: Partial<CSSStyleDeclaration>;
+}
+
+const ICON_HEIGHT = 16;
```
How can we enable scaling the icon with css in the use case where the user is changing the size for, say, mobile?

The rest of the button size with `--button-size`

ilyabo replied:
This size was only used for the viewBox, the icon can still be resized independently. But I removed the constant and hardcoded the values. I think it was not very useful, but rather confusing.

chrisgervang commented on modules/widgets/src/compass-widget.tsx:
```diff
@@ -0,0 +1,118 @@
+const ICON_HEIGHT = 16;
+
+class CompassWidget implements Widget<CompassWidgetProps> {
+  id = 'Compass';
```
`compass`

chrisgervang replied:
`compass`

chrisgervang commented on modules/widgets/src/compass-widget.tsx:
```diff
@@ -0,0 +1,118 @@
+  element?: HTMLDivElement;
+
+  constructor(props: CompassWidgetProps) {
+    this.id = props.id || 'Compass';
```
`compass`

chrisgervang commented on modules/widgets/src/compass-widget.tsx:
```diff
@@ -0,0 +1,118 @@
+  update() {
+    const [rz, rx] = this.getRotation();
+    const ui = (
+      <div className="deckgl-widget-button-group">
```
Can this be `deckgl-widget-button`?

chrisgervang commented on modules/widgets/src/stylesheet.css:
```diff
@@ -31,3 +31,24 @@
 }
+
+.deckgl-widget-button {
+  color: white;
```
The icon color is meant to them-able. Let's do `var(--icon-color, white)` and define the appropriate color in `themes.ts`. I'll do this in the fullscreen PR for you to pull in.

ilyabo replied:
I'll also add `--icon-compass-north-color` and `--icon-compass-south-color`

chrisgervang commented on modules/widgets/src/zoom-widget.tsx:
```diff
@@ -0,0 +1,108 @@
+  }
+}
+
+const Button = props => {
```
Can this be reused in compass? You can add a `components.tsx` for these reused things.

chrisgervang commented on examples/get-started/pure-js/widgets/app.js:
```diff
@@ -1,14 +1,20 @@
 
-const prefersDarkScheme = window.matchMedia("(prefers-color-scheme: dark)");
-const widgetTheme = prefersDarkScheme.matches ? DarkGlassTheme : LightGlassTheme
+const prefersDarkScheme = window.matchMedia('(prefers-color-scheme: dark)');
```
Odd, are you running `yarn lint fix` too on the root? I wonder why this differs

chrisgervang commented on modules/widgets/src/zoom-widget.tsx:
```diff
@@ -0,0 +1,114 @@
+    const ui = (
+      <div className="deckgl-widget-button-group">
+        <Button onClick={() => this.handleZoomIn()} label={this.props.zoomInLabel}>
+          <path d="M12 4.5v15m7.5-7.5h-15" />
```
I'd like to use rounded google symbols for icons. Can you please use these? You may want to modify how the fill is applied.

**Zoom In**
```
<svg xmlns="http://www.w3.org/2000/svg" width="28" height="28" viewBox="0 0 28 28" fill="none">
  <path d="M13.1029 14.8971H8.89332C8.64334 14.8971 8.43196 14.8105 8.25918 14.6373C8.08639 14.4641 8 14.2521 8 14.0014C8 13.7508 8.08639 13.5396 8.25918 13.368C8.43196 13.1963 8.64334 13.1105 8.89332 13.1105H13.1029V8.90097C13.1029 8.65161 13.1895 8.43912 13.3627 8.26348C13.5359 8.08783 13.7479 8 13.9986 8C14.2492 8 14.4604 8.08783 14.632 8.26348C14.8037 8.43912 14.8895 8.65161 14.8895 8.90097V13.1105H19.099C19.3484 13.1105 19.5609 13.1971 19.7365 13.3703C19.9122 13.5436 20 13.7555 20 14.0062C20 14.2569 19.9122 14.468 19.7365 14.6397C19.5609 14.8113 19.3484 14.8971 19.099 14.8971H14.8895V19.1067C14.8895 19.3567 14.8029 19.568 14.6297 19.7408C14.4564 19.9136 14.2445 20 13.9938 20C13.7431 20 13.532 19.9136 13.3603 19.7408C13.1887 19.568 13.1029 19.3567 13.1029 19.1067V14.8971Z" fill="#18181A"/>
</svg>
```

**Zoom Out**
```
<svg xmlns="http://www.w3.org/2000/svg" width="28" height="28" viewBox="0 0 28 28" fill="none">
  <path d="M8.89332 14.89C8.64334 14.89 8.43196 14.8032 8.25918 14.6296C8.08639 14.4561 8 14.2437 8 13.9926C8 13.7415 8.08639 13.5299 8.25918 13.3579C8.43196 13.186 8.64334 13.1 8.89332 13.1H19.099C19.349 13.1 19.5617 13.1868 19.737 13.3603C19.9123 13.5339 20 13.7462 20 13.9974C20 14.2485 19.9123 14.46 19.737 14.632C19.5617 14.804 19.349 14.89 19.099 14.89H8.89332Z" fill="#18181A"/>
</svg>
```

chrisgervang commented on modules/widgets/src/zoom-widget.tsx:
```diff
@@ -0,0 +1,114 @@
+    if (className) element.classList.add(className);
+    Object.entries(style).map(([key, value]) => element.style.setProperty(key, value as string));
+    const ui = (
+      <div className="deckgl-widget-button-group">
```
Can you please adjust this to have a 1px gap between the buttons? I think its fine to add new css classes for the group use case similar to `.deckgl-widget-button-border` since they can reuse the same variables where it makes sense, such as `--button-shadow`

chrisgervang commented on modules/widgets/src/compass-widget.tsx:
```diff
@@ -0,0 +1,112 @@
+  onAdd({deck}: {deck: Deck}): HTMLDivElement {
+    const {style, className} = this.props;
+    const element = document.createElement('div');
+    element.classList.add('deckgl-widget', 'deckgl-widget-compass');
```
After discussion with @ibgreen and @Pessimistress, we've decided to drop the `gl`. Couple reasons, it's shorter, we're long-term distancing from `gl` (onto WebGPU), and tooltip widget has always used `.deck-tooltip`.

chrisgervang commented:
I've merged #8024 so this can target master now.

ilyabo commented:
Thanks @chrisgervang! Will try to update it soon

chrisgervang commented on .eslintrc.cjs:
```diff
@@ -115,7 +115,8 @@ module.exports = getESLintConfig({
           // For widgets module. Disable React-style JSX linting since they conflict with Preact JSX.
-          'react/react-in-jsx-scope': 0
+          'react/react-in-jsx-scope': 0,
+          'no-unused-expressions': ['error', { 'allowShortCircuit': true }]
```
Added this to allow 
```
 handleZoomIn() {
    this.viewport && this.handleZoom(this.viewport.zoom + 1);
  }
```
Is this ok @Pessimistress? Or would we prefer another way to guard?

Pessimistress replied:
We use if everywhere else

ilyabo commented:
Thanks @chrisgervang for resolving the conflicts!

ilyabo commented:
> I've updated styles and icons to match our [Figma design](https://www.figma.com/file/KGQqLbZZGUINgyyKYPvDvu/Widgets?node-id=419%3A1602&mode=dev), and added a horizontal option for the zoom widget (or any button group). There is some polish to do, but this is pretty close.

Looks really cool, thank you!
